js/Ice/react-greeter/src/App.tsx

```diff
@@ -0,0 +1,127 @@
+import React, { useState, useCallback, useEffect, useRef } from "react";
+import { Ice } from "@zeroc/ice";
+import { VisitorCenter } from "./Greeter.js";
+
+import StatusMessage from "./StatusMessage";
+
+interface AppState {
+    name: string;
+    greeting: string;
+    loading: boolean;
+    error: string;
+}
+
+function App(): JSX.Element {
+    const [state, setState] = useState<AppState>({
+        name: "React",
+        greeting: "",
+        loading: false,
+        error: "",
+    });
+
+    // Refs to hold the communicator and greeter proxy
+    const communicatorRef = useRef<Ice.Communicator | null>(null);
+    const greeterRef = useRef<VisitorCenter.GreeterPrx | null>(null);
+
+    // Initialize communicator and greeter proxy when component mounts
+    useEffect(() => {
+        try {
+            // Create an Ice communicator
+            const communicator = Ice.initialize();
+            communicatorRef.current = communicator;
+
+            // Create a proxy to the Greeter object. If you run the server on a different computer, replace localhost
+            // with the server's hostname or IP address
+            const greeter = new VisitorCenter.GreeterPrx(communicator, "greeter:ws -h localhost -p 4061");
+            greeterRef.current = greeter;
+        } catch (error) {
+            console.error("Error initializing Ice:", error);
+            setState((prev) => ({
+                ...prev,
+                error: `Failed to initialize Ice: ${error instanceof Error ? error.message : "Unknown error"}`,
+            }));
+        }
+
+        return () => {
+            const cleanup = async () => {
+                try {
+                    if (communicatorRef.current) {
+                        await communicatorRef.current.destroy();
+                    }
+                } catch (err) {
+                    console.warn("Failed to destroy communicator:", err);
+                }
+            };
+            // Kick off the async cleanup, but don't return the promise
+            cleanup();
```

Comment:
Is this extra step necessary?

Reply:
Need to check, it is definitely ugly.

Reply:
The callback returned by useEffect must be not async

```
src/App.tsx:27:15 - error TS2345: Argument of type '() => () => Promise<void>' is not assignable to parameter of type 'EffectCallback'.
  Type '() => Promise<void>' is not assignable to type 'void | Destructor'.
    Type '() => Promise<void>' is not assignable to type 'Destructor'.
      Type 'Promise<void>' is not assignable to type 'void | { [UNDEFINED_VOID_ONLY]: never; }'.

27     useEffect(() => {            ~~~~~~~
```

If we call `destroy` directly and not await it we can end up with an unhalted promise rejection.

Reply:
Either one is fine with me.